### PR TITLE
Review

### PR DESCRIPTION
**Pull request by emeraldTree24**
All the files in the astroquery/cadc folder are what I've done. I copied the tap folder from astroquery/utils/tap.

### CODE REVIEW

File: astroquery/cadc/core.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+
+        Returns
+        -------
+        CadcTAP object
+        """
+        if tap_plus_handler is None:
+            if url is None:
+                self.__cadctap = TapPlus(
```
Comment (andamian) [inline astroquery/cadc/core.py]: double underscore not necessary. Simple suffices

File: astroquery/cadc/auth.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import sys
+import getpass
+
+
+class AuthMethod(object):
+    """
+    Class to be inherited by all the authentication method classes
+    """
+    def __init__(self, auth_type):
+        self.__authMethod = auth_type
```
Comment (andamian) [inline astroquery/cadc/auth.py]: Please document method

File: astroquery/cadc/auth.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        Initialize
+
+        Parameter
+        certificate : location of file of certificate to use
+                       for authentication
+        """
+        super(CertAuthMethod, self).__init__('certificate')
+        if certificate is not None:
+            assert certificate is not '' and os.path.isfile(certificate),\
```
Comment (andamian) [inline astroquery/cadc/auth.py]: In general, assert should not be used in production code (just test code) because it is removed when python is run in optimized mode. Just use if and raise.

File: astroquery/cadc/auth.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                   which looks in the home directory for a file
+                   called '.netrc'
+        """
+        super(NetrcAuthMethod, self).__init__('netrc')
+        self._username = username
+        self._hosts_auth = {}
+        if username is None:
+            hosts = netrclib.netrc(filename).hosts
+            self._netrc = os.path.join(os.environ['HOME'], ".netrc")
+            for host_name in hosts:
```
Comment (andamian) [inline astroquery/cadc/auth.py]: for or if?

File: astroquery/cadc/core.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                    url="http://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/tap",
```
Comment (andamian) [inline astroquery/cadc/core.py]: This should be a const declared at the beginning of the file.

File: astroquery/cadc/core.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                    verbose=verbose)
+            else:
+                self.__cadctap = TapPlus(url=url, verbose=verbose)
+        else:
+            self.__cadctap = tap_plus_handler
```
Comment (andamian) [inline astroquery/cadc/core.py:51]: Maybe raise an exception if url is set but, in this case, ignored?

File: astroquery/cadc/auth.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                self._hosts_auth[host_name] = (hosts[host_name][0],
+                                               hosts[host_name][2])
+        else:
+            self._netrc = None
+
+    def get_auth(self, realm):
+        """
+        Returns a user/password touple for the given realm.
+
+        :param realm: realm for the authentication
```
Comment (andamian) [inline astroquery/cadc/auth.py]: Inconsistent style for documentation. This is the Sphinx based style used in opencadc projects. Here please use the numpy style that you used for the other functions

File: astroquery/cadc/core.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+=============
+
+"""
+
+from astroquery.cadc.tap import TapPlus
+
+__all__ = ['Cadc', 'CadcTAP']
+
+
+class CadcTAP(object):
```
Comment (andamian) [inline astroquery/cadc/core.py:19]: This class adds no value. Why would a client that already has a tap_plus_handler would use it?
For the other cases you could do away with a simple function get_tap_handler(url, verbose) or CadcTAP could simply subclass TapPlus and set the default url in the __init__.

File: astroquery/cadc/tap/conn/tapconn.py [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+
+        Returns
+        -------
+        The suitable file extension based on the output format
+        """
+        if output_format is None:
+            return ".vot"
+        ext = ""
+        outputFormat = output_format.lower()
+        if "vot" in outputFormat:
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tapconn.py]: This is looks like a great candidate for a dictionary...

File: astroquery/cadc/tap/conn/tapconn.py [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+        Returns
+        -------
+        The suitable file extension based on the HTTP(s) headers
+        """
+        if headers is None:
+            return ""
+        ext = ""
+        contentType = self.find_header(headers, 'Content-Type')
+        if contentType is not None:
+            contentType = contentType.lower()
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tapconn.py]: Dictionary that could be used here maybe?

File: astroquery/cadc/tap/conn/tapconn.py [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+import mimetypes
+import time
+
+__all__ = ['TapConn']
+
+CONTENT_TYPE_POST_DEFAULT = "application/x-www-form-urlencoded"
+
+
+class TapConn(object):
+    """TAP plus connection class
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tapconn.py]: I assume that you used this from GAIA and added authentication, maybe? If it is a new implementation I suggest using the requests library instead. It could simply the class a lot.

File: astroquery/cadc/tap/conn/tapconn.py [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+class ConnectionHandler(object):
+    def __init__(self, host, port, sslport):
+        self.__connHost = host
+        self.__connPort = port
+        self.__connPortSsl = sslport
+
+    def get_connection(self, ishttps=False, verbose=False):
+        if ishttps:
+            if verbose:
+                print("------>https")
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tapconn.py]: libraries don't usually use print directly. They do logging.debug instead.

File: astroquery/cadc/tap/conn/tests/__init__.py [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tests/__init__.py]: File not needed as tests is not a package

File: astroquery/cadc/tap/conn/tests/DummyConnHandler.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+=============
+
+"""
+from astroquery.cadc.tap import taputils
+
+from astropy.extern.six.moves.urllib.parse import urlencode
+from astroquery.cadc.tap.xmlparser import utils
+
+
+class DummyConnHandler(object):
```
Comment (andamian) [inline astroquery/cadc/tap/conn/tests/DummyConnHandler.py]: Same thing could be achieved with Mock classes with less lines of code.

File: astroquery/cadc/tap/model/tests/__init__.py [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (andamian) [inline astroquery/cadc/tap/model/tests/__init__.py]: File not needed

**Review by andamian — CHANGES_REQUESTED** [review]

File: astroquery/cadc/cadctap/core.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+class TapPlusCadc(TapPlus):
+    """TAP plus class
+    Provides TAP and TAP+ capabilities
+    """
+
+    def __init__(self, url=None, host=None, server_context=None,
+                 tap_context=None, port=80, sslport=443,
+                 default_protocol_is_https=False, connhandler=None,
+                 verbose=True):
+        """Constructor
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py:43]: Please explain the purpose. What is doing that the parent class couldn't.

File: astroquery/cadc/cadctap/core.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+
+__all__ = ['TapPlusCadc']
+
+VERSION = "1.0.1"
+TAP_CLIENT_ID = "aqtappy-" + VERSION
+
+
+class TapPlusCadc(TapPlus):
+    """TAP plus class
+    Provides TAP and TAP+ capabilities
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py:32]: Please explain the purpose.

File: astroquery/cadc/cadctap/core.py [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+                                          port,
+                                          port)
+        else:
+            connHandler = TapConnCadc(default_protocol_is_https,
+                                      host,
+                                      server_context,
+                                      tap_context,
+                                      port,
+                                      sslport)
+        if connhandler is not None:
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: I would move this at the beginning and avoid doing unnecessary work in 'if url' if connhandler is passed by the client.

File: astroquery/cadc/cadctap/core.py [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+        ----------
+        table : str, mandatory
+            full qualified table name (i.e. schema name + table name)
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+        Returns
+        -------
+        A table object
+        """
+        print("Retrieving table '"+str(table)+"'")
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: Not a good idea to use print for debugging messages because users have no way of turning them off. Use logging instead or if you want to keep it consistent with the host package, check the verbose flag before printing.

File: astroquery/cadc/cadctap/core.py [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+                                        include_shared_tables=False,
+                                        verbose=verbose)
+        for tab in tables:
+            if tab.get_qualified_name() == table:
+                return tab
+        return None
+
+    def _Tap__launchJobMultipart(self, query, uploadResource, uploadTableName,
+                                 outputFormat, context, verbose, name=None):
+        uploadValue = str(uploadTableName) + ",param:" + str(uploadTableName)
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py:140]: Why did you have to override this method?

File: astroquery/cadc/core.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+            job
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+        """
+        return self.__cadctap.save_results(job, filename, verbose)
+
+    def login(self, user=None, password=None,
+              certificate_file=None, verbose=False):
+        """
+        Saves job results
```
Comment (andamian) [inline astroquery/cadc/core.py]: Wrong documentation

File: astroquery/cadc/core.py [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+
+        return self.__cadctap.login(user=user, password=password,
+                                    certificate_file=certificate_file,
+                                    verbose=verbose)
+
+    def logout(self, verbose=False):
+        return self.__cadctap.logout(verbose)
+
+
+Cadc = CadcTAP()
```
Comment (andamian) [inline astroquery/cadc/core.py:255]: What is this for?

File: astroquery/cadc/cadctap/core.py [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+            location of the certificate
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+        """
+        if certificate_file is None and user is None:
+            print('Input user/password or certificate file path')
+            self._TapPlus__isLoggedIn = True
+            return
+        if certificate_file is not None and user is not None:
+            print('Choose one form of authentication only')
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: Typically these cases translate into raising AttributeError. How can the caller figure out what is going on? Do they need to parse the stdout? Comment applies to all the error cases.

File: astroquery/cadc/cadctap/core.py [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+        if user is not None and password is not None:
+            if self.__certificate is not None:
+                print('Already logged in with certificate')
+                return
+            self._TapPlus__user = user
+            self._TapPlus__pwd = password
+            self.__dologin(verbose)
+        else:
+            if password is None:
+                print("Invalid password")
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: Rather no password at all

File: astroquery/cadc/cadctap/core.py [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+        response = self.__execLogin(self._TapPlus__user,
+                                    self._TapPlus__pwd,
+                                    verbose)
+        # check response
+        connHandler = self._TapPlus__getconnhandler()
+        isError = connHandler.check_launch_response_status(response,
+                                                           verbose,
+                                                           200)
+        if isError:
+            print("Login error: " + str(response.reason))
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: print without verbose. I would probably explore the use of logging and replace all/most of the print statements in the package.

File: astroquery/cadc/cadctap/core.py [hunk 26]
```diff
@@ -0,0 +1,339 @@
...
+            flag to display information about the process
+        """
+        self._TapPlus__getconnhandler()._TapConn__connectionHandler. \
+            _ConnectionHandlerCadc__certificate = None
+        if self._TapPlus__getconnhandler().cookies_set():
+            self._TapPlus__getconnhandler().unset_cookie()
+        self.__certificate = None
+        self._TapPlus__isLoggedIn = False
+
+    def __execLogin(self, usr, pwd, verbose=False):
```
Comment (andamian) [inline astroquery/cadc/cadctap/core.py]: Rename to execCadcLogin or add argument with the login url.

File: astroquery/cadc/cadctap/job.py [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+import requests
+import astroquery.cadc.cadctap.jobSaxParser
+from astroquery.utils.tap.model.job import Job
+from astroquery.cadc.cadctap import utils
+
+__all__ = ['Job']
+
+
+class JobCadc(Job):
+    """Job class
```
Comment (andamian) [inline astroquery/cadc/cadctap/job.py]: Comment the reason the original Job class was not sufficient.

File: astroquery/cadc/cadctap/job.py [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+                self.wait_for_job_end(verbose)
+                context = 'async/' + str(self.jobid) + "/results/result"
+                response = self.connHandler.execute_get(context)
+                if verbose:
+                    print(response.status, response.reason)
+                    print(response.getheaders())
+
+                numberOfRedirects = 0
+                while (response.status == 303 or response.status == 302) and \
+                        numberOfRedirects < 20:
```
Comment (andamian) [inline astroquery/cadc/cadctap/job.py]: what is this numberOfRedirects limit and what is it trying to fix?

File: astroquery/cadc/cadctap/job.py [hunk 29]
```diff
@@ -0,0 +1,212 @@
...
+
+            lphase = responseData.lower().strip()
+            if verbose:
+                print("Job " + self.jobid + " status: " + lphase)
+            if "pending" != lphase and "queued" != lphase and \
+                    "executing" != lphase:
+                break
+            # PENDING, QUEUED, EXECUTING, COMPLETED, ERROR, ABORTED, UNKNOWN,
+            # HELD, SUSPENDED, ARCHIVED:
+            time.sleep(0.5)
```
Comment (andamian) [inline astroquery/cadc/cadctap/job.py]: I think that the wait time is too short. I would probably do a few tries at 1s for short queries and then the rest at 30s.
To be complete, you might want to trap here the Key interrupts and abort the job.

File: astroquery/cadc/cadctap/job.py [hunk 30]
```diff
@@ -0,0 +1,212 @@
...
+                raise requests.exceptions.HTTPError(
+                    'Error running query, PHASE: '+wjData)
+        subContext = "async/" + str(self.jobid) + "/results/result"
+        resultsResponse = self.connHandler.execute_get(subContext)
+        if debug:
+            print(resultsResponse.status, resultsResponse.reason)
+            print(resultsResponse.getheaders())
+
+        numberOfRedirects = 0
+        while (resultsResponse.status == 303 or resultsResponse.status == 302)\
```
Comment (andamian) [inline astroquery/cadc/cadctap/job.py:206]: This looks similar to save_results. Can common code be factored out?

File: astroquery/cadc/cadctap/jobSaxParser.py [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+UWS_MESSAGE = "uws:message"
+
+VALID_ITEMS = [UWS_JOBID, UWS_RUNID, UWS_OWNERID, UWS_PHASE, UWS_QUOTE,
+               UWS_START_TIME, UWS_END_TIME, UWS_CREATION_TIME,
+               UWS_EXECUTION_DURATION, UWS_DESTRUCTION, UWS_LOCATIONID,
+               UWS_NAME, UWS_PARAMETER, UWS_MESSAGE]
+
+
+class JobSaxParserCadc(JobSaxParser):
+    '''
```
Comment (andamian) [inline astroquery/cadc/cadctap/jobSaxParser.py:34]: Again, not clear why is this needed...

File: astroquery/cadc/cadctap/tests/DummyConnHandler.py [hunk 32]
```diff
@@ -0,0 +1,56 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+from astroquery.utils.tap.conn.tests.DummyConnHandler import DummyConnHandler
+
+
+class DummyConnHandlerCadc(DummyConnHandler):
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/DummyConnHandler.py:6]: An alternative to this is to mock the connection handler in the test.

File: astroquery/cadc/cadctap/tests/__init__.py [hunk 33]
```diff
@@ -0,0 +1,7 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/__init__.py:1]: Not needed: tests is not a package

File: astroquery/cadc/cadctap/tests/data/cert.pem [hunk 34]
```diff
@@ -0,0 +1 @@
+fake certificate for testing
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/data/cert.pem]: Most of the time these, very simple files, are not needed as they can be either faked with mock patch or created dynamically in a temp dir.

File: astroquery/cadc/cadctap/tests/setup_package.py [hunk 35]
```diff
@@ -0,0 +1,16 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import absolute_import
+
+import os
+
+
+# setup paths to the test data
+# can specify a single file or a list of files
+def get_package_data():
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/setup_package.py:9]: Is this required

File: astroquery/cadc/cadctap/tests/test_tap.py [hunk 36]
```diff
@@ -0,0 +1,569 @@
...
+
+        jobDataFile = data_path('job_1.vot')
+        jobData = utils.read_file_content(jobDataFile)
+        responseRedirect.set_data(method='GET',
+                                  context=None,
+                                  body=jobData,
+                                  headers=None)
+        req = "http://test:1111/tap/async/" + str(jobid) + "/redirect"
+        connHandler.set_response(req, responseRedirect)
+        tap.save_results(job, 'file.txt')
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/test_tap.py:474]: Test with no asserts? Have a missed something?

File: astroquery/cadc/cadctap/tests/test_tap.py [hunk 37]
```diff
@@ -0,0 +1,569 @@
...
+        tap = TapPlusCadc("http://test:1111/tap", connhandler=connHandler)
+        tap.logout()
+        assert tap._TapPlus__user is None, \
+            "User was set"
+        assert tap._TapPlus__pwd is None, \
+            "Password was set"
+        assert tap._TapPlusCadc__certificate is None, \
+            "Certificate was set"
+
+    def __find_column(self, columnName, columns):
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/test_tap.py:543]: simple preceding _ is sufficient.

File: astroquery/cadc/cadctap/tests/test_tap.py [hunk 38]
```diff
@@ -0,0 +1,569 @@
...
+from astroquery.utils.tap.xmlparser import utils
+from astroquery.utils.tap import taputils
+
+
+def data_path(filename):
+    data_dir = os.path.join(os.path.dirname(__file__), 'data')
+    return os.path.join(data_dir, filename)
+
+
+class TestTapCadc(unittest.TestCase):
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/test_tap.py:22]: You could have used just pytest and not unittest, in which case this class would have been unnecessary.

File: astroquery/cadc/cadctap/tests/test_tap.py [hunk 39]
```diff
@@ -0,0 +1,569 @@
...
+            (column.name, dataType, column.data_type)
+        assert column.flag == flag, \
+            "Wrong flag for table %s. Expected: '%s', found '%s'" % \
+            (column.name, flag, column.flag)
+
+    def __check_results_column(self, results, columnName, description, unit,
+                               dataType):
+        c = results[columnName]
+        assert c.description == description, \
+            "Wrong description for results column '%s'. " \
```
Comment (andamian) [inline astroquery/cadc/cadctap/tests/test_tap.py:568]: general comment: unless you want to be consistent with the current package, I would have used the '{}'.format string syntax instead.

File: astroquery/cadc/cadctap/utils.py [hunk 40]
```diff
@@ -0,0 +1,61 @@
...
+=============
+
+"""
+import io
+from astropy import units as u
+from astropy.table import Table as APTable
+import six
+
+
+def util_create_string_from_buffer(buffer):
```
Comment (andamian) [inline astroquery/cadc/cadctap/utils.py:21]: Don't understand this. Is this about serializing a list? ''.join(list) should do the trick, no? As for different types of string representations in python 2 and 3, we deal with that transparently with:
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

File: astroquery/cadc/cadctap/utils.py [hunk 41]
```diff
@@ -0,0 +1,61 @@
...
+                        col.unit.name.replace(".", " ").replace("'", ""))
+                except Exception as ex:
+                    pass
+            elif isinstance(col.unit, str):
+                col.unit = col.unit.replace(".", " ").replace("'", "")
+
+    return result
+
+
+def get_suitable_astropy_format(outputFormat):
```
Comment (andamian) [inline astroquery/cadc/cadctap/utils.py:56]: Dictionaries are the Pythonic ways to do this mapping.

File: astroquery/cadc/cadctap/utils.py [hunk 42]
```diff
@@ -0,0 +1,61 @@
...
+    if "csv" == outputFormat:
+        return "ascii.csv"
+    if "tsv" == outputFormat:
+        return "ascii.fast_tab"
+    return outputFormat
+
+
+def read_file_content(filePath):
```
Comment (andamian) [inline astroquery/cadc/cadctap/utils.py:70]: Is this required, given that it's a 2 liner:
with open(file,'r') as f:
    output = f.read()

File: astroquery/cadc/tests/DummyJob.py [hunk 43]
```diff
@@ -0,0 +1,30 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+=============
+Cadc TAP plus
+=============
+
+"""
+
+
+class DummyJob(object):
```
Comment (andamian) [inline astroquery/cadc/tests/DummyJob.py:10]: Same comment as above: try to use Mock next time.

File: astroquery/cadc/tests/DummyTapHandler.py [hunk 44]
```diff
@@ -0,0 +1,121 @@
...
+=============
+Cadc TAP plus
+=============
+
+"""
+
+from astroquery.cadc.tests.DummyJob import DummyJob
+
+
+class DummyTapHandler(object):
```
Comment (andamian) [inline astroquery/cadc/tests/DummyTapHandler.py:12]: It seems like the implementation of Mock.

File: astroquery/cadc/tests/__init__.py [hunk 45]
```diff
@@ -0,0 +1,7 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (andamian) [inline astroquery/cadc/tests/__init__.py:1]: Remove file

**Review by andamian — CHANGES_REQUESTED** [review]
General comment is that documentation does not reveal what is being overridden from the tap library and what the reason was.

File: astroquery/cadc/cadctap/core.py (hunk 17 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py:32]: Done

File: astroquery/cadc/cadctap/core.py (hunk 16 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py:43]: Done

File: astroquery/cadc/cadctap/core.py (hunk 18 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: Done

File: astroquery/cadc/cadctap/core.py (hunk 19 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: I added a if verbose but any of the code it still uses from utils/tap is going to have print statments without checking if it's verbose

File: astroquery/cadc/cadctap/core.py (hunk 20 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py:140]: Added doc, because of separating creating and running the job

File: astroquery/cadc/cadctap/core.py (hunk 23 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: Made them raise AttributeError

File: astroquery/cadc/cadctap/core.py (hunk 24 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: Done

File: astroquery/cadc/cadctap/core.py (hunk 25 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: I could replace with logging but nothing else it's relying on is using it so I think it would be pointless

File: astroquery/cadc/cadctap/core.py (hunk 26 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/core.py]: I added a login_url argument with a default above

File: astroquery/cadc/cadctap/job.py (hunk 27 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/job.py]: done

File: astroquery/cadc/cadctap/job.py (hunk 28 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/job.py]: Just to make sure it doesn't do an infinite loop and hopefully it's less than 20 redirects away

File: astroquery/cadc/cadctap/job.py (hunk 29 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/job.py]: 0.5 is seconds so half a second, I added a loops variable to change the wait time and added a keyboard interrupt

File: astroquery/cadc/cadctap/jobSaxParser.py (hunk 31 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/jobSaxParser.py:34]: Done

File: astroquery/cadc/cadctap/tests/DummyConnHandler.py (hunk 32 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/tests/DummyConnHandler.py:6]: The documentation for the testing mentions this way of testing so that's why I used it and instead of writing a new way of testing I tried to use most of what was there

File: astroquery/cadc/cadctap/tests/__init__.py (hunk 33 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/tests/__init__.py:1]: The template for astroquery has one so in case they require it I put it there

File: astroquery/cadc/cadctap/tests/setup_package.py (hunk 35 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/tests/setup_package.py:9]: It gets all the files from the tests/data folder that aren't .py files

File: astroquery/cadc/cadctap/utils.py [hunk 46]
```diff
@@ -0,0 +1,61 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (emeraldTree24) [inline astroquery/cadc/cadctap/utils.py:1]: For this whole file I copied it and changed get_suitable_astropy_format() function. I couldn't figure out how to do the inheritance thing when it was just functions not classes and I don't want to change anything else in case it was needed that way for something

File: astroquery/cadc/core.py (hunk 8 above)
Comment (emeraldTree24) [reply astroquery/cadc/core.py:19]: Probably, I did it this way because that's how gaia did it and now there are a couple of function that have code in them to get everything to work right

File: astroquery/cadc/core.py (hunk 21 above)
Comment (emeraldTree24) [reply astroquery/cadc/core.py]: Done

File: astroquery/cadc/core.py (hunk 22 above)
Comment (emeraldTree24) [reply astroquery/cadc/core.py:255]: So you can import the object initialized, astroquery wants you to be able to do that

File: astroquery/cadc/tests/DummyJob.py (hunk 43 above)
Comment (emeraldTree24) [reply astroquery/cadc/tests/DummyJob.py:10]: I think astroquery likes it that way

File: astroquery/cadc/tests/DummyTapHandler.py (hunk 44 above)
Comment (emeraldTree24) [reply astroquery/cadc/tests/DummyTapHandler.py:12]: Making it like astroquery

File: astroquery/cadc/tests/__init__.py (hunk 45 above)
Comment (emeraldTree24) [reply astroquery/cadc/tests/__init__.py:1]: astroquery template

File: astroquery/cadc/cadctap/job.py (hunk 30 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/job.py:206]: Done

File: astroquery/cadc/cadctap/tests/data/cert.pem (hunk 34 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/tests/data/cert.pem]: I deleted this file and made it a tempfile

File: astroquery/cadc/core.py (hunk 6 above)
Comment (emeraldTree24) [reply astroquery/cadc/core.py:51]: I added a raise error if both are sent

File: astroquery/cadc/cadctap/job.py [hunk 47]
```diff
@@ -144,25 +117,63 @@ def wait_for_job_end(self, verbose=False):
...
+                lphase = responseData.lower().strip()
+                if verbose:
+                    print("Job " + self.jobid + " status: " + lphase)
+                if "pending" != lphase and "queued" != lphase and \
+                        "executing" != lphase:
+                    break
+                # PENDING, QUEUED, EXECUTING, COMPLETED, ERROR, ABORTED,
+                # UNKNOWN, HELD, SUSPENDED, ARCHIVED:
+                if loops < 10:
+                    seconds = 0.5
```
Comment (andamian) [inline astroquery/cadc/cadctap/job.py]: Make it 1s please. 0.5 is too little for an Internet connection...

File: astroquery/cadc/cadctap/jobSaxParser.py [hunk 48]
```diff
@@ -33,17 +33,30 @@
 class JobSaxParserCadc(JobSaxParser):
     '''
     classdocs
+    Reason for change
+    -----------------
+    Return JobCada insteasd of Job object
```
Comment (andamian) [inline astroquery/cadc/cadctap/jobSaxParser.py]: Cadc -> Cadc

File: astroquery/cadc/cadctap/job.py (hunk 47 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/job.py]: done

File: astroquery/cadc/cadctap/jobSaxParser.py (hunk 48 above)
Comment (emeraldTree24) [reply astroquery/cadc/cadctap/jobSaxParser.py]: done
